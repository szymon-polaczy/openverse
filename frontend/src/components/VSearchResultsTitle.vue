<template>
  <h1
    class="sr-only break-words md:not-sr-only"
    :class="[
      size === 'large'
        ? 'heading-2 lg:heading-1 !leading-none lg:!leading-none'
        : 'heading-2 !leading-none',
    ]"
  >
    <span aria-hidden="true">{{ searchTerm }}</span
    ><span class="sr-only">{{ ariaHeading }}</span>
  </h1>
</template>

<script lang="ts">
import { useI18n } from "#imports"

import { defineComponent, PropType, computed } from "vue"

import type { SupportedMediaType, SupportedSearchType } from "~/constants/media"

import { useGetLocaleFormattedNumber } from "~/composables/use-get-locale-formatted-number"
import { getCountKey } from "~/composables/use-i18n-utilities"

export default defineComponent({
  name: "VSearchResultsTitle",
  props: {
    searchTerm: {
      required: true,
      type: String,
    },
    size: {
      required: false,
      default: "default",
      type: String as PropType<"default" | "large">,
    },
    resultCounts: {
      required: true,
      type: Array as PropType<[SupportedMediaType, number][]>,
    },
    searchType: {
      required: true,
      type: String as PropType<SupportedSearchType>,
    },
  },
  setup(props) {
    const getLocaleFormattedNumber = useGetLocaleFormattedNumber()
    const i18n = useI18n()
    const mediaLocaleCounts = computed(() =>
      props.resultCounts.reduce(
        (acc, [mediaType, count]) => {
          return {
            ...acc,
            [mediaType]: {
              count,
              countKey: getCountKey(count),
              localeCount: getLocaleFormattedNumber(count),
            },
          }
        },
        {} as Record<
          SupportedMediaType,
          { count: number; countKey: string; localeCount: string }
        >
      )
    )

    const _getAllMediaAriaHeading = () => {
      const imageLocaleCounts = mediaLocaleCounts.value.image
      const imageResults = i18n.t(
        `browsePage.aria.allResultsHeadingCount.image.${imageLocaleCounts.countKey}`,
        {
          localeCount: imageLocaleCounts.localeCount,
        }
      )

      const audioLocaleCounts = mediaLocaleCounts.value.audio
      const audioResults = i18n.t(
        `browsePage.aria.allResultsHeadingCount.audio.${audioLocaleCounts.countKey}`,
        {
          localeCount: audioLocaleCounts.localeCount,
        }
      )

      return i18n
        .t("browsePage.aria.results.all", {
          query: props.searchTerm,
          imageResults,
          audioResults,
        })
        .toString()
    }

    const ariaHeading = computed((): string => {
      switch (props.searchType) {
        case "image": {
          const { count, countKey, localeCount } = mediaLocaleCounts.value.image
          return i18n.t(`browsePage.aria.results.image.${countKey}`, {
            count,
            localeCount,
            query: props.searchTerm,
          })
        }
        case "audio": {
          const { count, countKey, localeCount } = mediaLocaleCounts.value.audio
          return i18n.t(`browsePage.aria.results.audio.${countKey}`, {
            count,
            localeCount,
            query: props.searchTerm,
          })
        }
        default:
        case "all": {
          return _getAllMediaAriaHeading()
        }
      }
    })

    return {
      ariaHeading,
    }
  },
})
</script>

<style scoped>
/* Title case the user's search query */
h1:first-letter {
  text-transform: uppercase;
}
</style>
