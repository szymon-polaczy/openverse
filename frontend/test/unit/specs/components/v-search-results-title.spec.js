import { describe, expect, it } from "vitest"

import { render } from "~~/test/unit/test-utils/render"

import VSearchResultsTitle from "~/components/VSearchResultsTitle.vue"

const DEFAULT_PROPS = {
  size: "large",
  searchTerm: "zack",
  searchType: "all",
  resultCounts: [
    ["image", 10],
    ["audio", 10],
  ],
}

/**
 * Build test scenarios out of all possible
 * result count combinations.
 */
function getScenarios() {
  const counts = [0, 10, 4300, 10000, 10001]
  let scenarios = []
  counts.forEach((count) => {
    counts.forEach((innerCount) => {
      scenarios.push({
        searchType: "all",
        resultCounts: [
          ["image", count],
          ["audio", innerCount],
        ],
      })
    })
    scenarios.push({
      searchType: "image",
      resultCounts: [
        ["image", count],
        ["audio", 0],
      ],
    })
    scenarios.push({
      searchType: "audio",
      resultCounts: [
        ["audio", count],
        ["image", 0],
      ],
    })
  })
  return scenarios.map((scenario) => [JSON.stringify(scenario), scenario])
}

describe("VSearchResultsTitle", () => {
  const getOptions = ({ props, ...options } = {}) => ({
    ...options,
    props: {
      ...DEFAULT_PROPS,
      ...props,
    },
  })

  it("should render an h1 tag containing the correct text", async () => {
    const { container } = render(VSearchResultsTitle, getOptions())
    expect(container).toMatchSnapshot()
  })

  describe("accessible heading", () => {
    it.each(getScenarios())("%s", (_, { searchType, resultCounts }) => {
      const { container } = render(
        VSearchResultsTitle,
        getOptions({
          props: {
            searchType,
            resultCounts,
          },
        })
      )

      // Cannot use inline snapshot in `each` test
      expect(container).toMatchSnapshot()
    })
  })
})
