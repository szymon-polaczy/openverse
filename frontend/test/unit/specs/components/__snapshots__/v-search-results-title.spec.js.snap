// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",0],["audio",0]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",0],["audio",10]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",0],["audio",4300]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",0],["audio",10000]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",0],["audio",10001]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",10],["audio",0]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",10],["audio",10]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",10],["audio",4300]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",10],["audio",10000]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",10],["audio",10001]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",4300],["audio",0]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",4300],["audio",10]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",4300],["audio",4300]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",4300],["audio",10000]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",4300],["audio",10001]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",10000],["audio",0]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",10000],["audio",10]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",10000],["audio",4300]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",10000],["audio",10000]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",10000],["audio",10001]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",10001],["audio",0]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",10001],["audio",10]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",10001],["audio",4300]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",10001],["audio",10000]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"all","resultCounts":[["image",10001],["audio",10001]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"audio","resultCounts":[["audio",0],["image",0]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"audio","resultCounts":[["audio",10],["image",0]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"audio","resultCounts":[["audio",4300],["image",0]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"audio","resultCounts":[["audio",10000],["image",0]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"audio","resultCounts":[["audio",10001],["image",0]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"image","resultCounts":[["image",0],["audio",0]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"image","resultCounts":[["image",10],["audio",0]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"image","resultCounts":[["image",4300],["audio",0]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"image","resultCounts":[["image",10000],["audio",0]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > accessible heading > {"searchType":"image","resultCounts":[["image",10001],["audio",0]]} 1`] = `undefined`;

exports[`VSearchResultsTitle > should render an h1 tag containing the correct text 1`] = `undefined`;
